src/com/amazon/corretto/crypto/provider/HkdfSecretKeyFactorySpi.java

```diff
@@ -0,0 +1,168 @@
+
+import static com.amazon.corretto.crypto.provider.HkdfSpec.HKDF_EXPAND_MODE;
+import static com.amazon.corretto.crypto.provider.HkdfSpec.HKDF_EXTRACT_MODE;
+import static com.amazon.corretto.crypto.provider.HkdfSpec.HKDF_MODE;
+
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactorySpi;
+import javax.crypto.spec.SecretKeySpec;
+
+class HkdfSecretKeyFactorySpi extends SecretKeyFactorySpi {
+  private final int digestCode;
+  private final int digestLength;
+
+  private HkdfSecretKeyFactorySpi(final int digestCode, final int digestLength) {
+    this.digestCode = digestCode;
+    this.digestLength = digestLength;
+  }
+
+  @Override
+  protected SecretKey engineGenerateSecret(final KeySpec keySpec) throws InvalidKeySpecException {
+    if (!(keySpec instanceof HkdfSpec)) {
+      throw new InvalidKeySpecException("KeySpec must be an instance of HkdfSpec");
+    }
+    final HkdfSpec spec = (HkdfSpec) keySpec;
+
+    final byte[] resultBytes;
+    switch (spec.mode) {
+      case HKDF_MODE:
+        checkExpandLength(spec.desiredLength);
+        resultBytes = new byte[spec.desiredLength];
+        hkdf(
+            resultBytes,
+            resultBytes.length,
+            digestCode,
+            spec.secretOrPrk,
+            spec.secretOrPrk.length,
+            spec.salt,
+            spec.salt.length,
+            spec.info,
+            spec.info.length);
+        break;
+      case HKDF_EXTRACT_MODE:
+        // In extract mode, we ignore the value of desiredLength and use
+        // the output size of the digest as desiredLength
+        resultBytes = new byte[digestLength];
+        hkdfExtract(
+            resultBytes,
+            resultBytes.length,
+            digestCode,
+            spec.secretOrPrk,
+            spec.secretOrPrk.length,
+            spec.salt,
+            spec.salt.length);
+        break;
+      case HKDF_EXPAND_MODE:
+        checkExpandLength(spec.desiredLength);
+        resultBytes = new byte[spec.desiredLength];
+        hkdfExpand(
+            resultBytes,
+            resultBytes.length,
+            digestCode,
+            spec.secretOrPrk,
+            spec.secretOrPrk.length,
+            spec.info,
+            spec.info.length);
+        break;
+      default:
+        throw new AssertionError(
+            "This should not be reachable, since the constructor of HkdfSpec ensures the value is"
+                + " one of the above choices.");
+    }
+
+    return new SecretKeySpec(resultBytes, spec.algorithmName);
+  }
+
+  private void checkExpandLength(final long outLen) {
+    final long digestLengthLong = digestLength;
+    final long maxInt = Integer.MAX_VALUE;
+    final long n = (outLen + digestLengthLong - 1L) / digestLengthLong;
+    if (outLen + digestLengthLong > maxInt || n > 255) {
```

Comment:
Since we are using longs, in case of overflow, `outLen + digestLengthLong > maxInt` becomes true.
LC has a similar logic: https://github.com/aws/aws-lc/blob/161e747063dcb16e71d299190f072a26c0a56294/crypto/fipsmodule/hkdf/hkdf.c#L94-L99.

I think `outLen >= digestLenghtLong * 255L`  is sufficient as you pointed out. I'll update the PR with that.